packages/flutter_tools/lib/src/runner/target_devices.dart

```diff
@@ -302,3 +338,397 @@ class TargetDevices {
+    }
+
+    // Wait for user to select a device.
+    chosenDevice = await futureChosenDevice;
+
+    // Update the [DeviceManager.specifiedDeviceId] so that the user will not
+    // be prompted again.
+    _deviceManager.specifiedDeviceId = chosenDevice.id;
+
+    return <Device>[chosenDevice];
+  }
+
+  /// Reprint list of attached devices before printing list of wireless devices.
+  Future<void> _verbosePrintWirelessDevices(
+    List<Device> attachedDevices,
+    List<Device> wirelessDevices,
+  ) async {
+    if (wirelessDevices.isEmpty) {
+      _logger.printStatus(_noWirelessDevicesFoundMessage);
+    }
+    // The iOS xcdevice outputs once wireless devices are done loading, so
+    // reprint attached devices so they're grouped with the wireless ones.
+    _logger.printStatus(_connectedDevicesMessage);
+    await Device.printDevices(attachedDevices, _logger);
+
+    if (wirelessDevices.isNotEmpty) {
+      _logger.printStatus('');
+      _logger.printStatus(_wirelesslyConnectedDevicesMessage);
+      await Device.printDevices(wirelessDevices, _logger);
+    }
+  }
+
+  /// Clear [numLinesToClear] lines from terminal. Print message and list of
+  /// wireless devices.
+  Future<void> _printWirelessDevices(
+    List<Device> wirelessDevices,
+    int numLinesToClear,
+  ) async {
+    _logger.printStatus(
+      globals.terminal.clearLines(numLinesToClear),
+      newline: false,
+    );
+    _logger.printStatus('');
+    if (wirelessDevices.isEmpty) {
+      _logger.printStatus(_noWirelessDevicesFoundMessage);
+    } else {
+      _logger.printStatus(_wirelesslyConnectedDevicesMessage);
+      await Device.printDevices(wirelessDevices, _logger);
+    }
+  }
+}
+
+@visibleForTesting
+class TargetDeviceSelection {
+  TargetDeviceSelection(this._logger);
+
+  List<Device> devices = <Device>[];
+  final Logger _logger;
+  int invalidAttempts = 0;
+
+  /// Prompt user to select a device and wait until they select a valid device.
+  ///
+  /// If the user selects `q`, exit the tool.
+  ///
+  /// If the user selects an invalid number, reprompt them and continue waiting.
+  Future<Device> userSelectDevice() async {
+    Device? chosenDevice;
+    while (chosenDevice == null) {
+      final String userInputString = await readUserInput();
+      if (userInputString.toLowerCase() == 'q') {
+        throwToolExit('');
+      }
+      final int deviceIndex = int.parse(userInputString) - 1;
+      if (deviceIndex < devices.length) {
+        chosenDevice = devices[deviceIndex];
+      }
+    }
+
+    return chosenDevice;
+  }
+
+  /// Prompt user to select a device and wait until they select a valid
+  /// character.
+  ///
+  /// Only allow input of a number or `q`.
+  @visibleForTesting
+  Future<String> readUserInput() async {
+    final RegExp pattern = RegExp(r'\d+$|q', caseSensitive: false);
+    final String prompt = userMessages.flutterChooseOne;
+    String? choice;
+    globals.terminal.singleCharMode = true;
```

Comment:
Can you use `promptForCharInput` instead?

Reply:
:+1:

Looks like you'll have to check first if the user is in an interactive session (unless the callers of `readUserInput` already guard that).

Reply:
Not without changing it extensively. `promptForCharInput` requires a list of acceptable characters and then only displays and allows input of those characters. With the new process, more devices might be loaded while `promptForCharInput` is still waiting for user input and it can't be updated to use a new list of acceptable characters.

Reply:
this is fine